### ISSUE TITLE
High memory usage inference RTMP stream

### ISSUE THREAD

**Issue by dcarrion87** [ultralytics/ultralytics#8144]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Simillar to https://github.com/ultralytics/ultralytics/issues/3623 but it's now closed.

We're seeing the same issue with RTMP stream. We're looking at ways to make inference as lean as possible. Model was trained using pretrained yolov8 nano and only uses 4 classes. Loading model takes 400MB RAM but as soon as prediction starts blows out to 2.5-3.5GB which is way too much for lean implementation we're trying to achieve. We're using it like this:

```
    results = model.predict(source=STREAM, verbose=False,show=False,conf=0.70, imgsz=[736,1280], stream=True)
    for r in results:
      # Skip processing for testing
      continue
      # Processing happens here
``` 

![Screenshot 2024-02-12 at 6 18 38 pm](https://github.com/ultralytics/ultralytics/assets/2899172/8ffbbb1e-6802-4303-abd5-b0c8cdef8881)

The memory utilisation actually explodes before the stream loop. It's on that model.predict line.

Should we be considering different model if we're looking for something less resource intensive at inference?

**Comment by glenn-jocher** [comment]
@dcarrion87 hi there! 👋 It looks like you're experiencing high memory usage during RTMP stream inference. To help reduce memory consumption, you can try the following:

1. Ensure you're using the latest version of YOLOv8, as we continuously optimize for performance.
2. If you haven't already, consider using the `half` precision inference by setting `half=True` in your `predict` call, which can significantly reduce memory usage on compatible GPUs.
3. Reduce the `imgsz` if possible, as smaller images require less memory.
4. If you're not using the `stream_buffer`, set it to `False` to avoid buffering all frames.

Here's an example with these adjustments:
```python
results = model.predict(source=STREAM, verbose=False, show=False, conf=0.70, imgsz=[736,1280], half=True, stream=True, stream_buffer=False)
```

If memory usage is still higher than expected, you might want to consider a custom lightweight model tailored for your specific use case. Let us know how it goes, and we're here to help with any further optimizations! 😊

**Comment by dcarrion87** [comment]
Thanks @glenn-jocher. Unfortunately no difference. Memory still high and we've discovered CPU usage is insanely high even  despite GPU usage. I think we're going to have to go a different route potentially with more low level language to make this optimal.

**Comment by dcarrion87** [comment]
Also just a note this seems to happen just with single frames/images as well. There's something not right with the way memory is being used / allocated.

**Comment by glenn-jocher** [comment]
@dcarrion87 the memory usage you're experiencing during inference on an RTMP stream may be due to several factors, including the size of the model, the image resolution being processed, and the batch size. Since you're already using YOLOv8 nano, which is one of the smaller models, here are a few suggestions to reduce memory usage:

1. **Reduce Image Size:** The `imgsz` parameter defines the input image size for the model. You're currently using `[736,1280]`, which is relatively large. Try reducing the image size to see if it significantly impacts accuracy. Smaller images will use less memory during processing.

2. **Stream Buffering:** Ensure that the `stream_buffer` parameter is set to `False` to avoid buffering all streaming frames. This will help in reducing memory usage by processing only the most recent frame.

3. **Batch Size:** If you're processing frames in batches, consider reducing the batch size. However, since you're using a stream, this may not be applicable unless you're batching frames before processing.

4. **Half Precision:** Using half precision (`FP16`) can reduce memory usage. Set the `half` parameter to `True` if it's not already.

5. **Disable Unnecessary Features:** If you're not using certain features like visualization (`show`), ensure they are disabled to save memory.

6. **Garbage Collection:** In some cases, Python's garbage collector might not be releasing memory promptly. You can manually invoke garbage collection after processing each frame or batch of frames using `gc.collect()`.

7. **Model Quantization:** Consider quantizing your model to reduce its size and memory footprint. This involves converting the model weights to lower precision (e.g., `INT8`), which can lead to a smaller model with faster inference times and lower memory usage.

8. **Optimize Model for Inference:** If you haven't already, optimize your model for inference using techniques like pruning and knowledge distillation to reduce its size and complexity without significantly impacting accuracy.

9. **System Monitoring:** Monitor your system's memory usage to identify if there are any other processes that might be consuming a significant amount of memory, which could be affecting the available memory for your application.

10. **Different Model Architecture:** If the above steps do not yield the desired memory usage reduction, you may need to consider using a different model architecture that is designed to be more memory-efficient.

Remember that reducing memory usage may come at the cost of reduced accuracy or increased inference time. You'll need to find a balance that meets your requirements for both performance and resource utilization.

**Comment by dcarrion87** [comment]
No I don't think so. I'm starting to realise this is drivers / CUDA being the culprit. The fact that not using CUDA resolves the issue. I'm also seeing it in a C++ implementation I've done. How can it possibly use that much memory in libraries!

**Comment by glenn-jocher** [comment]
@dcarrion87 the memory usage you're experiencing during RTMP stream inference with YOLOv8 may be influenced by several factors, including the size of the model, the resolution of the input frames, and the batch size used during inference. Since you're aiming for a lean implementation, here are a few suggestions to help reduce memory usage:

1. **Reduce Input Resolution**: The `imgsz` parameter defines the input image size for the model. Reducing the resolution can significantly decrease memory usage, but it may also affect the accuracy of the model. You're currently using an input size of `[736,1280]`. Try reducing it further if the accuracy remains acceptable for your use case.

2. **Stream Buffering**: Ensure that the `stream_buffer` parameter is set to `False` to avoid buffering all streaming frames in memory. This will help in reducing memory usage by only keeping the most recent frame in memory.

3. **Model Size**: YOLOv8 nano is already a small model, but you can try using an even smaller custom model if you're looking for less resource-intensive inference. This may involve training a new model with fewer layers or channels.

4. **Optimize Model**: If you haven't already, consider exporting the model to an optimized format like ONNX or TensorRT, which can help reduce memory usage during inference.

5. **Batch Size**: If you're processing frames one by one, ensure that the batch size is set to 1 to minimize memory usage.

6. **Automatic Mixed Precision (AMP)**: If your hardware supports it, using AMP can reduce memory usage by performing certain operations in half-precision.

7. **Garbage Collection**: In your processing loop, explicitly invoke Python's garbage collection after processing each frame to ensure that any unused memory is freed up.

Here's an example of how you might adjust your code:

```python
import gc
from ultralytics import YOLO

# Load the model with reduced input size and AMP enabled
model = YOLO('path/to/yolov8n.pt', imgsz=[736, 1280], half=True)

# Predict with stream=True and stream_buffer=False
results = model.predict(source=STREAM, verbose=False, show=False, conf=0.70, stream=True, stream_buffer=False)

for r in results:
    # Skip processing for testing
    continue
    # Processing happens here
    # ...

    # Explicitly invoke garbage collection
    gc.collect()
```

If you're still experiencing high memory usage after trying these suggestions, it may be worth opening a new issue or discussion on the Ultralytics YOLOv8 GitHub repository to get further assistance from the maintainers and community.

**Comment by dcarrion87** [comment]
As mentioned it seems related to NVIDIA drivers/libraries being loaded into memory.

**Comment by glenn-jocher** [comment]
@dcarrion87 the memory usage you're experiencing during inference on an RTMP stream with YOLOv8 seems to be higher than what you're aiming for. Here are a few suggestions to help reduce memory usage:

1. **Reduce Input Size**: The `imgsz` parameter defines the input size of the images that the model will process. Smaller image sizes can significantly reduce memory usage, but they may also affect the accuracy of the model. You're currently using an input size of `[736,1280]`. Try reducing it further if accuracy is not severely impacted.

2. **Use Half Precision**: If you're running on a GPU that supports half precision (FP16), you can enable it by setting the `half` parameter to `True`. This can reduce memory usage and potentially increase inference speed.

3. **Stream Buffer**: Ensure that the `stream_buffer` parameter is set to `False`. This prevents buffering all streaming frames and instead returns the most recent frame, which can help manage memory usage.

4. **Batch Size**: If you're processing frames in batches, consider reducing the batch size. However, since you're processing a stream, this may not be applicable.

5. **Model Selection**: If reducing input size and enabling half precision do not sufficiently reduce memory usage, you may need to consider using a more lightweight model architecture. YOLOv8 Nano is already a smaller model, but you might need to look into custom lightweight models or other architectures that are designed for edge devices and have lower memory footprints.

6. **Optimize Model**: If you haven't already, you can try exporting the model to an optimized format like ONNX or TensorRT, which can sometimes offer better memory and speed performance during inference.

7. **Garbage Collection**: Ensure that any unused variables or data are being properly garbage collected. In Python, you can force garbage collection by importing the `gc` module and running `gc.collect()`.

8. **System Monitoring**: Use system monitoring tools to check if there are any memory leaks or if other processes on the system are consuming a significant amount of memory.

9. **Custom Inference Loop**: Instead of using the built-in `predict` method, you could write a custom inference loop that more tightly controls memory allocation and deallocation, although this would require a deeper understanding of the YOLOv8 internals.

If you've tried all the above and are still facing high memory usage, it might be worth opening a new issue or discussion on the Ultralytics YOLOv8 GitHub repository to get direct support from the maintainers. They might have more insights or optimizations specific to your use case.

**Comment by Wang-Yu-Qing** [comment]
GPT answers made this place really bad.

**Comment by glenn-jocher** [comment]
@Wang-Yu-Qing hi there! 👋 We're sorry to hear about your concerns. If you're facing any specific issues with YOLOv8 or have suggestions for improvement, we'd love to hear them! Your feedback is valuable in helping us enhance the experience for everyone. 🚀

**Comment by ooolky** [comment]
@dcarrion87 Have you tried upgrading your CUDA version? I’m using an RTX 4070 for inference, and when I upgraded my CUDA version from 11.8 to 12.4, RAM usage for each YOLO model decreased from 3.5GB to 1.8GB.
